core/src/type_info.rs

```diff
@@ -0,0 +1,304 @@
+            "SQL_DATETIME_SUB".to_string(),
+            Schema::Atomic(Atomic::Scalar(BsonTypeName::Int)),
+            Nullability::NULLABLE
+        ),
+        MongoColMetadata::new(
+            "",
+            "".to_string(),
+            "NUM_PREC_RADIX".to_string(),
+            Schema::Atomic(Atomic::Scalar(BsonTypeName::Int)),
+            Nullability::NULLABLE
+        ),
+        MongoColMetadata::new(
+            "",
+            "".to_string(),
+            "INTERVAL_PRECISION".to_string(),
+            Schema::Atomic(Atomic::Scalar(BsonTypeName::Int)),
+            Nullability::NULLABLE
+        ),
+    ];
+}
+
+#[derive(Debug)]
+pub struct MongoTypesInfo {
+    current_type_index: isize,
+    sql_data_type: SqlDataType,
+}
+
+impl MongoTypesInfo {
+    pub fn new(sql_data_type: SqlDataType) -> MongoTypesInfo {
+        MongoTypesInfo {
+            current_type_index: 0,
+            sql_data_type,
+        }
+    }
+}
+
+impl MongoStatement for MongoTypesInfo {
+    // iterate through the list, searching for the next "valid" data type.
+    // a type is valid if its sql type matches the desired sql type, or if we are getting all types.
+    fn next(&mut self, _: Option<&MongoConnection>) -> Result<bool> {
+        loop {
+            self.current_type_index += 1;
+            if self.current_type_index > (DATA_TYPES.len() as isize)
+                || DATA_TYPES[(self.current_type_index - 1) as usize].sql_type == self.sql_data_type
+                || self.sql_data_type == SqlDataType(0)
+            {
+                break;
+            }
+        }
+        Ok(self.current_type_index <= (DATA_TYPES.len() as isize))
+    }
+
+    // Get the BSON value for the cell at the given colIndex on the current row.
+    fn get_value(&self, col_index: u16) -> Result<Option<Bson>> {
+        // 1 -> TYPE_NAME
+        // 2 -> DATA_TYPE
+        // 3 -> COLUMN_SIZE
+        // 4 -> LITERAL_PREFIX
+        // 5 -> LITERAL_SUFFIX
+        // 6 -> CREATE_PARAMS
+        // 7 -> NULLABLE
+        // 8 -> CASE_SENSITIVE
+        // 9 -> SEARCHABLE
+        // 10 -> UNSIGNED_ATTRIBUTE
+        // 11 -> FIXED_PREC_SCALE
+        // 12 -> AUTO_UNIQUE_VALUE
+        // 13 -> LOCAL_TYPE_NAME
+        // 14 -> MINIMUM_SCALE
+        // 15 -> MAXIMUM_SCALE
+        // 16 -> SQL_DATA_TYPE
+        // 17 -> SQL_DATETIME_SUB
+        // 18 -> NUM_PREC_RADIX
+        // 19 -> INTERVAL_PRECISION
+        match DATA_TYPES.get((self.current_type_index - 1) as usize) {
+            Some(type_info) => {
+                Ok(Some(match col_index {
+                    1 | 13 => Bson::String(type_info.type_name.to_string()),
+                    2 | 16 => Bson::Int32(type_info.sql_type.0 as i32),
+                    3 => match type_info.precision {
+                        Some(precision) => Bson::Int32(precision as i32),
+                        None => Bson::Null,
+                    },
+                    4 | 5 => match **type_info {
```

Comment:
For all of these that match `**type_info` itself (`4`, `5`, `8`, `10`, `12`, `17`, and `18`), I wonder if we should make these optional fields on the `BsonTypeInfo` struct so that we could follow the same pattern as the others! `match type_info.<relevant_field> { Some(val) => val | None => Null }`. What do you think?

Reply:
I think this makes a lot of sense, especially given Natacha's feedback as well. Will look into this 😁